### PR TITLE
Formalize the Ping Protocol

### PR DESCRIPTION
**Pull request by sappenin**
This PR has been dormant for a very long time, but I'm reintroducing it for a few reasons:

1. We have a new [proposals process,](https://github.com/interledger/rfcs/blob/master/CONTRIBUTING.md#proposals) and formalizing "ping" still seems like a good proposal.
1. This has been on my TODO list forever
1. Uni-directional mode seems useful in "send-only" modes (see discussion [here](https://github.com/interledgerjs/ilp-protocol-stream/issues/137))
1. Bi-directional mode is deployed in the JS connector and also seems useful in certain production networks.

Links for Context

* Interledger forum discussion ([here](https://forum.interledger.org/t/deprecate-the-echo-protocol-and-replace-with-one-way-ping/520/5)).
* Previous PR from long ago ([here](https://github.com/interledger/rfcs/pull/516)).

### CODE REVIEW

**Comment by sappenin** [comment]
> I am unsure what mechanism the receiver would know that this is specifically a PING packet

In unidirectional mode, the receiver knows this is a `ping` packet based upon two things: 1) The packet is destined to the Connector's operating address; 2) The packet uses the well-known condition. 

In my mind, there's no data payload needed in unidirectional flow. In bi-directional flow, the data payload is needed to signal receiver where to send the `pong` packets to.

File: proposals/0000-ilp-ping-protocol/0000-ilp-ping-protocol.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+## Scope
+It is already possible to test end-to-end connectivity in Interledger by sending a small payment using existing mechanisms. However, doing so requires higher-level protocols, such as [SPSP](https://github.com/interledger/rfcs/blob/master/0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md) and [STREAM](https://github.com/interledger/rfcs/blob/master/0029-stream/0029-stream.md).
+
+Instead, this protocol uses only [Interledger-layer](https://github.com/interledger/rfcs/blob/master/0001-interledger-architecture/0001-interledger-architecture.md#interledger-protocol) technologies to test connectivity between two nodes in the Interledger without requiring any additional protocols or outside information.
+ 
+This protocol is not designed to debug routing issues and does not provide additional diagnostic information about the state of routing. That use case would be better served by a separate `traceroute` mechanism.
+
+All Interledger implementations _SHOULD_ respond to ping protocol requests, unless a node has a specific reason not to, such as security or privacy concerns.
```
Comment (sharafian) [inline proposals/0000-ilp-ping-protocol/0000-ilp-ping-protocol.md:15]: can we clarify implementation here? Is a STREAM server expected to respond to pings, or just connectors?

Comment (sappenin) [reply proposals/0000-ilp-ping-protocol/0000-ilp-ping-protocol.md:15]: hmmm...great question. I'm not sure I have a strong opinion. In Java-land, we've moved most of our receiver logic into the Connector runtime, so this deployment style would sort of get ping functionality for free.

I suppose after re-reading line 15 above, it's sort of optional for a STREAM receiver to support this functionality -- maybe that's enough?

File: proposals/0000-ilp-ping-protocol/images/bidirectional-flow.svg [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+          </g>
+          <g id="Group_7">
+            <title>Graphic_57</title>
+            <g id="Graphic_9">
+              <rect x="107.65458" y="326.16667" width="82.01172" height="14" fill="white"/>
+            </g>
+            <g id="Graphic_8">
+              <title>Text</title>
+              <text transform="translate(112.65458 326.66667)" fill="black">
+                <tspan font-family="Courier" font-size="12" font-weight="400" fill="black" x="0" y="11">(4)Fulfill</tspan>
```
[the remark this answers is not in the document]
Comment (sappenin) [reply proposals/0000-ilp-ping-protocol/images/bidirectional-flow.svg]: I updated the images to make them more flow-based - take another look?

**Comment by matdehaast** [comment]
> > I am unsure what mechanism the receiver would know that this is specifically a PING packet
> 
> In unidirectional mode, the receiver knows this is a `ping` packet based upon two things: 1) The packet is destined to the Connector's operating address; 2) The packet uses the well-known condition.
> 
> In my mind, there's no data payload needed in unidirectional flow. In bi-directional flow, the data payload is needed to signal receiver where to send the `pong` packets to.

Thanks that makes sense.

File: proposals/0000-ilp-ping-protocol/0000-ilp-ping-protocol.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+This RFC specifies a similar mechanism for Interledger nodes to test `uptime`, `response-time` and approximate `path cost` for a particular Initiator/Recipient pair.
+
+## Scope
+It is already possible to test end-to-end connectivity in Interledger by sending a small payment using existing mechanisms. However, doing so requires higher-level protocols, such as [SPSP](https://github.com/interledger/rfcs/blob/master/0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md) and [STREAM](https://github.com/interledger/rfcs/blob/master/0029-stream/0029-stream.md).
+
+Instead, this protocol uses only [Interledger-layer](https://github.com/interledger/rfcs/blob/master/0001-interledger-architecture/0001-interledger-architecture.md#interledger-protocol) technologies to test connectivity between two nodes in the Interledger without requiring any additional protocols or outside information.
```
Comment (kincaidoneil) [inline proposals/0000-ilp-ping-protocol/0000-ilp-ping-protocol.md:11]: > Instead, this protocol uses only [Interledger-layer](https://github.com/interledger/rfcs/blob/master/0001-interledger-architecture/0001-interledger-architecture.md#interledger-protocol) technologies

Nit: the ping protocol itself _is_ a higher layer protocol on top of ILP 😉

**Pull request closed without merging**